src/helpers/pull-request-parsing.ts

```diff
@@ -0,0 +1,261 @@
+  if (includedFileDiffs.length === 0) {
+    logger.error(`Cannot include any files from diff without exceeding token limits.`);
+    return { diff: null };
+  }
+
+  // Accurately calculate token count for included files we have approximated to be under the limit
+  const accurateFileDiffStats = await Promise.all(
+    includedFileDiffs.map(async (file) => {
+      const tokenCountArray = await encodeAsync(file.diffContent, { disallowedSpecial: new Set() });
+      const tokenCount = tokenCountArray.length;
+      return { ...file, tokenCount };
+    })
+  );
+
+  // Take an accurate reading of our current collection of files within the diff
+  currentTokenCount = accurateFileDiffStats.reduce((sum, file) => sum + file.tokenCount, runningTokenCount);
+
+  // Remove files from the end of the list until we are within token limits
+  while (currentTokenCount > tokensRemaining && accurateFileDiffStats.length > 0) {
+    const removedFile = accurateFileDiffStats.pop();
+    currentTokenCount -= removedFile?.tokenCount || 0;
+    logger.info(`Excluded ${removedFile?.filename || "Unknown filename"} after accurate token count exceeded limits.`);
+  }
+
+  if (accurateFileDiffStats.length === 0) {
+    logger.error(`Cannot include any files from diff after accurate token count calculation.`);
+    return { diff: null };
+  }
+
+  // Build the diff with the included files
+  const currentDiff = accurateFileDiffStats.map((file) => file.diffContent).join("\n");
+
+  return { diff: currentDiff };
+}
+
+// Helper to speed up tokenization
+export async function encodeAsync(text: string, options: EncodeOptions): Promise<number[]> {
+  return new Promise((resolve) => {
+    const result = encode(text, options);
+    resolve(result);
+  });
+}
+
+// Helper to parse a diff into per-file diffs
+export function parsePerFileDiffs(diff: string): { filename: string; diffContent: string }[] {
+  // regex to capture diff sections, including the last file
+  const diffPattern = /^diff --git a\/(.*?) b\/.*$/gm;
+  let match: RegExpExecArray | null;
+  const perFileDiffs = [];
+  let lastIndex = 0;
+
+  // iterate over each file in the diff
+  while ((match = diffPattern.exec(diff)) !== null) {
+    const filename = match[1];
+    const startIndex = match.index;
+
+    // if we have pushed a file into the array, "append" the diff content
+    if (perFileDiffs.length > 0) {
+      perFileDiffs[perFileDiffs.length - 1].diffContent = diff.substring(lastIndex, startIndex).trim();
+    }
+
+    perFileDiffs.push({ filename, diffContent: "" });
+    lastIndex = startIndex;
+  }
+  // append the last file's diff content
+  if (perFileDiffs.length > 0 && lastIndex < diff.length) {
+    perFileDiffs[perFileDiffs.length - 1].diffContent = diff.substring(lastIndex).trim();
+  }
+
+  return perFileDiffs;
+}
+
+// This speeds things up considerably by skipping non-readable/non-relevant files
+function isEssentialFile(filename: string): boolean {
```

Comment:
Black listing is definitely the wrong approach. I explained a dynamic strategy elsewhere that will exclude the most changes to the least.

Reply:
@UbiquityOS explain the strategy I mentioned

Reply:
I like to test and see how it performs wherever I can but I haven't been satisfied with the results lately.

Reply:
i implemented your suggested strategy mate this is a move towards improve exec speed by ignoring files outright that are not readable

Reply:
removed it but if you think it's a good idea we can always return it